lib/reversetunnel/remoteclientmanager.go

```diff
@@ -0,0 +1,308 @@
+	accessPoint auth.RemoteProxyAccessPoint
+	nodeWatcher *services.NodeWatcher
+	caWatcher   *services.CertAuthorityWatcher
+}
+
+func newRemoteClientManager(ctx context.Context, config remoteClientManagerConfig) (*remoteClientManager, error) {
+	if config.newClientFunc == nil {
+		return nil, trace.BadParameter("missing new client func")
+	}
+	if config.newAccessPointFunc == nil {
+		return nil, trace.BadParameter("missing new access point func")
+	}
+	if config.newNodeWatcherFunc == nil {
+		return nil, trace.BadParameter("missing new node watcher func")
+	}
+	if config.newCAWatcher == nil {
+		return nil, trace.BadParameter("missing new ca watcher func")
+	}
+	if config.log == nil {
+		return nil, trace.BadParameter("missing logger")
+	}
+
+	m := &remoteClientManager{
+		remoteClientManagerConfig: config,
+		parentCtx:                 ctx,
+		connectedOnceOrClosed:     make(chan struct{}),
+	}
+	return m, nil
+}
+
+// connect runs until all clients have been successfully setup.
+func (m *remoteClientManager) connect(ctx context.Context) error {
+	var (
+		client      auth.ClientI
+		accessPoint auth.RemoteProxyAccessPoint
+		nodeWatcher *services.NodeWatcher
+		caWatcher   *services.CertAuthorityWatcher
+		err         error
+	)
+
+	firstIteration := true
+	for {
+		if err := ctx.Err(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		// Use backoff after first attempt.
+		if !firstIteration {
+			<-m.backoff.After()
+			m.backoff.Inc()
+		} else {
+			firstIteration = false
+		}
+
+		if client == nil {
+			client, err = m.newClientFunc(ctx)
+			if err != nil {
+				m.log.WithError(err).Warnf("Failed to connect to remote auth server.")
+				continue
+			}
+		}
+
+		response, err := client.Ping(ctx)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to get remote auth server version.")
+			continue
+		}
+
+		version := response.ServerVersion
+		if accessPoint != nil {
+			accessPoint.Close()
+		}
+		accessPoint, err = m.newAccessPointFunc(ctx, client, version)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote access point.")
+			continue
+		}
+
+		if nodeWatcher != nil {
+			nodeWatcher.Close()
+		}
+		nodeWatcher, err = m.newNodeWatcherFunc(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote node watcher.")
+			continue
+		}
+
+		if caWatcher != nil {
+			caWatcher.Close()
+		}
+		caWatcher, err = m.newCAWatcher(ctx, accessPoint)
+		if err != nil {
+			m.log.WithError(err).Warnf("Failed to create remote CA watcher.")
+			continue
+		}
+
+		m.clientLock.Lock()
+		m.auth = client
+		m.accessPoint = accessPoint
+		m.nodeWatcher = nodeWatcher
+		m.caWatcher = caWatcher
+		m.once.Do(func() {
+			close(m.connectedOnceOrClosed)
+		})
+		m.clientLock.Unlock()
+
+		return nil
+	}
+}
+
+// Connect blocks while all clients are created. If connect is called more than once
+// existing clients will be closed and new clients created. Connect will always fail
+// after Close is called.
```

Comment:
What's the point of being able to `Connect` multiple times? We only call it once from `newRemoteSite`, and it makes things really confusing.

Reply:
I started writing this thinking we would want to reconnect. But now realize the underlying clients should reconnect on their own. Let me refactor this and it should address some of your concerns on it being too complicated. 😄